wasm/derive/src/lib.rs

```diff
@@ -31,6 +31,9 @@ pub fn iroha_wasm(_: TokenStream, item: TokenStream) -> TokenStream {
     quote! {
         #[no_mangle]
         unsafe extern "C" fn _iroha_wasm_main(ptr: u32, len: u32) {
+            #[cfg(debug_assertions)]
+            debug_assert!(false, "Only smartcontracts built with `release` profile are allowed");
```

Comment:
This is the best I could do to ensure only optimized builds are allowed. Malicious actor could still bypass this if they were to write smartcontract on their own and not via this crate. I don't think it's possible to determine a build is optimized from wasm blob

Reply:
Fees would be an incentive to optimize.
I don't know how to know the computation cost in advance though

Reply:
You can't check if something is built in release mode per se, release mode is really just `strip` + optimisations; but you can check the WASM executable for the presence of debug symbols, which would catch 
```
[profile.release]
debug = true
```
as a debug build. 

With that said, it definitely deserves its own issue, and a separate PR.

Reply:
#1866